SystemIncludesExtractor stripping `-stdlib` with `-query-driver`

Very similar to #1410 but for a different flag - the `-stdlib=libc++` flag to `clang` affects the system include location, but is stripped by `SystemIncludesExtractor` when using `query-driver`. I have a patch which keeps the `stdlib` flag in https://github.com/llvm/llvm-project/compare/main...cjc25:llvm-project:clangd-stdlib-support. Let me know if there are any reasons _not_ to send it as a pull request!

--

I built `clang`, `clangd`, `libc++abi`, and `lib++` from llvm-project/llvm commit eab5d337f0f62828a991ad7ed7e4257735c48e11. The following is a pretty small repro:

```
cjc25@host:~/src/bug-report$ cat bug-repro.cc 
#include <semaphore> // On my machine, present for clang libc++, but not libstdc++.
int main(int argc, char** argv) {
    std::counting_semaphore<> c{2};
    c.acquire();
    return 0;
}
cjc25@host:~/src/bug-report$ ../llvm/build/bin/clang -x c++ -stdlib=libstdc++ -lstdc++ -rpath ../llvm/build/lib/x86_64-unknown-linux-gnu/ -- bug-repro.cc
bug-repro.cc:1:10: fatal error: 'semaphore' file not found
    1 | #include <semaphore> // On my machine, present for clang libc++, but not libstdc++.
      |          ^~~~~~~~~~~
1 error generated.
cjc25@host:~/src/bug-report$ ../llvm/build/bin/clang -x c++ -stdlib=libc++ -lc++ -rpath ../llvm/build/lib/x86_64-unknown-linux-gnu/ -- bug-repro.cc
cjc25@host:~/src/bug-report$ echo $?
0
```

With an unmodified clangd:
```
cjc25@host:~/src/bug-report$ ../llvm/build/bin/clangd --query-driver=** --check=bug-repro.cc
I[21:41:16.022] clangd version 18.0.0 (https://github.com/llvm/llvm-project eab5d337f0f62828a991ad7ed7e4257735c48e11)
I[21:41:16.022] Features: linux
I[21:41:16.022] PID: 29892
I[21:41:16.022] Working directory: /home/cjc25/src/bug-report
I[21:41:16.022] argv[0]: ../llvm/build/bin/clangd
I[21:41:16.022] argv[1]: --query-driver=**
I[21:41:16.022] argv[2]: --check=bug-repro.cc
I[21:41:16.022] Entering check mode (no LSP server)
I[21:41:16.022] Testing on source file /home/cjc25/src/bug-report/bug-repro.cc
I[21:41:16.022] Loading compilation database...
I[21:41:16.023] Loaded compilation database from /home/cjc25/src/bug-report/compile_commands.json
I[21:41:16.053] System includes extractor: successfully executed /home/cjc25/src/bug-report/../llvm/build/bin/clang
        got includes: "/usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10, /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/x86_64-linux-gnu/c++/10, /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/backward, /usr/local/include, /usr/include/x86_64-linux-gnu, /usr/include"
        got target: "x86_64-unknown-linux-gnu"
I[21:41:16.053] Compile command from CDB is: [/home/cjc25/src/bug-report] ../llvm/build/bin/clang -stdlib=libc++ -lc++ -rpath ../llvm/build/lib/x86_64-unknown-linux-gnu/ -isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10 -isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/x86_64-linux-gnu/c++/10 -isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/backward -isystem /usr/local/include -isystem /usr/include/x86_64-linux-gnu -isystem /usr/include --target=x86_64-unknown-linux-gnu -resource-dir=/home/cjc25/src/llvm/build/lib/clang/18 -- /home/cjc25/src/bug-report/bug-repro.cc
I[21:41:16.053] Parsing command...
I[21:41:16.054] internal (cc1) args are: -cc1 -triple x86_64-unknown-linux-gnu -fsyntax-only -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name bug-repro.cc -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/home/cjc25/src/bug-report -fcoverage-compilation-dir=/home/cjc25/src/bug-report -resource-dir /home/cjc25/src/llvm/build/lib/clang/18 -isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10 -isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/x86_64-linux-gnu/c++/10 -isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/backward -isystem /usr/local/include -isystem /usr/include/x86_64-linux-gnu -isystem /usr/include -internal-isystem ../llvm/build/bin/../include/x86_64-unknown-linux-gnu/c++/v1 -internal-isystem ../llvm/build/bin/../include/c++/v1 -internal-isystem /home/cjc25/src/llvm/build/lib/clang/18/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -no-round-trip-args -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -x c++ /home/cjc25/src/bug-report/bug-repro.cc
I[21:41:16.054] Building preamble...
I[21:41:16.187] Built preamble of size 2876640 for file /home/cjc25/src/bug-report/bug-repro.cc version null in 0.13 seconds
I[21:41:16.187] Indexing headers...
E[21:41:16.230] [ambiguous_reference] Line 1: in included file: reference to 'integral_constant' is ambiguous
E[21:41:16.230] [fatal_too_many_errors] Line 1: too many errors emitted, stopping now
I[21:41:16.230] Building AST...
I[21:41:16.233] Indexing AST...
I[21:41:16.233] Building inlay hints
I[21:41:16.233] Building semantic highlighting
I[21:41:16.233] Testing features at each token (may be slow in large files)
I[21:41:16.236] Found definition heuristically in index for acquire
I[21:41:16.236] Found definition heuristically in index for acquire
I[21:41:16.236] All checks completed, 2 errors
```

With a patched clangd:

```
cjc25@host:~/src/bug-report$ ../llvm/build/bin/clangd --query-driver=** --check=bug-repro.cc
I[22:32:38.249] clangd version 18.0.0 (https://github.com/llvm/llvm-project bbb699071d9a4be996fcdf0dcb605fb42e74c0e6)
I[22:32:38.249] Features: linux
I[22:32:38.249] PID: 19073
I[22:32:38.249] Working directory: /home/cjc25/src/bug-report
I[22:32:38.249] argv[0]: ../llvm/build/bin/clangd
I[22:32:38.249] argv[1]: --query-driver=**
I[22:32:38.249] argv[2]: --check=bug-repro.cc
I[22:32:38.249] Entering check mode (no LSP server)
I[22:32:38.249] Testing on source file /home/cjc25/src/bug-report/bug-repro.cc
I[22:32:38.249] Loading compilation database...
I[22:32:38.250] Loaded compilation database from /home/cjc25/src/bug-report/compile_commands.json
I[22:32:38.280] System includes extractor: successfully executed /home/cjc25/src/bug-report/../llvm/build/bin/clang
        got includes: "/home/cjc25/src/llvm/build/bin/../include/x86_64-unknown-linux-gnu/c++/v1, /home/cjc25/src/llvm/build/bin/../include/c++/v1, /usr/local/include, /usr/include/x86_64-linux-gnu, /usr/include"
        got target: "x86_64-unknown-linux-gnu"
I[22:32:38.280] Compile command from CDB is: [/home/cjc25/src/bug-report] ../llvm/build/bin/clang -stdlib=libc++ -lc++ -rpath ../llvm/build/lib/x86_64-unknown-linux-gnu/ -isystem /home/cjc25/src/llvm/build/bin/../include/x86_64-unknown-linux-gnu/c++/v1 -isystem /home/cjc25/src/llvm/build/bin/../include/c++/v1 -isystem /usr/local/include -isystem /usr/include/x86_64-linux-gnu -isystem /usr/include --target=x86_64-unknown-linux-gnu -resource-dir=/home/cjc25/src/llvm/build/lib/clang/18 -- /home/cjc25/src/bug-report/bug-repro.cc
I[22:32:38.280] Parsing command...
I[22:32:38.281] internal (cc1) args are: -cc1 -triple x86_64-unknown-linux-gnu -fsyntax-only -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name bug-repro.cc -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fdebug-compilation-dir=/home/cjc25/src/bug-report -fcoverage-compilation-dir=/home/cjc25/src/bug-report -resource-dir /home/cjc25/src/llvm/build/lib/clang/18 -isystem /home/cjc25/src/llvm/build/bin/../include/x86_64-unknown-linux-gnu/c++/v1 -isystem /home/cjc25/src/llvm/build/bin/../include/c++/v1 -isystem /usr/local/include -isystem /usr/include/x86_64-linux-gnu -isystem /usr/include -internal-isystem ../llvm/build/bin/../include/x86_64-unknown-linux-gnu/c++/v1 -internal-isystem ../llvm/build/bin/../include/c++/v1 -internal-isystem /home/cjc25/src/llvm/build/lib/clang/18/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/10/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -fdeprecated-macro -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -no-round-trip-args -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -x c++ /home/cjc25/src/bug-report/bug-repro.cc
I[22:32:38.281] Building preamble...
I[22:32:38.404] Built preamble of size 2914372 for file /home/cjc25/src/bug-report/bug-repro.cc version null in 0.12 seconds
I[22:32:38.404] Indexing headers...
I[22:32:38.439] Building AST...
I[22:32:38.443] Indexing AST...
I[22:32:38.443] Building inlay hints
I[22:32:38.443] Building semantic highlighting
I[22:32:38.443] Testing features at each token (may be slow in large files)
I[22:32:38.447] All checks completed, 0 errors
```

Note that while this check command reports ambiguous headers, in LSP via VSCode it appears not to find the libc++ headers at all (LSP reporting that `<semaphore>` was missing, despite a working build which matched `compile_commands.json`, was how I found this issue in the first place).

I can force check to look the same as LSP by modifying the `compiler_commands.json` to exclude `-stdlib=libc++` (which doesn't compile). Then it reports the missing `<semaphore>` header I expected. I don't fully understand where check observes `-stdlib=libc++` to add headers in addition to the ones queried from the driver. It looks like it'd be where check builds the CC1 invocation, but I can't find where it'd add new system headers. Teaching query-driver to pass `-stdlib` gets rid of the error in both the check and LSP, but if someone already knows where those -internal-isystem arguments are coming from I'd love to learn before merging this patch!

--- a/clang-tools-extra/clangd/SystemIncludeExtractor.cpp
+++ b/clang-tools-extra/clangd/SystemIncludeExtractor.cpp
@@ -88,12 +88,14 @@ struct DriverArgs {
   std::string Sysroot;
   std::string ISysroot;
   std::string Target;
+  std::string Stdlib;
 
   bool operator==(const DriverArgs &RHS) const {
     return std::tie(Driver, StandardIncludes, StandardCXXIncludes, Lang,
-                    Sysroot, ISysroot, Target) ==
+                    Sysroot, ISysroot, Target, Stdlib) ==
            std::tie(RHS.Driver, RHS.StandardIncludes, RHS.StandardCXXIncludes,
-                    RHS.Lang, RHS.Sysroot, RHS.ISysroot, RHS.Target);
+                    RHS.Lang, RHS.Sysroot, RHS.ISysroot, RHS.Target,
+                    RHS.Stdlib);
   }
 
   DriverArgs(const tooling::CompileCommand &Cmd, llvm::StringRef File) {
@@ -136,6 +138,13 @@ struct DriverArgs {
       } else if (Arg.consume_front("-target")) {
         if (Arg.empty() && I + 1 < E)
           Target = Cmd.CommandLine[I + 1];
+      } else if (Arg.consume_front("--stdlib")) {
+        if (Arg.consume_front("="))
+          Stdlib = Arg.str();
+        else if (Arg.empty() && I + 1 < E)
+          Stdlib = Cmd.CommandLine[I + 1];
+      } else if (Arg.consume_front("-stdlib=")) {
+        Stdlib = Arg.str();
       }
     }
 
@@ -175,6 +184,8 @@ struct DriverArgs {
       Args.append({"-isysroot", ISysroot});
     if (!Target.empty())
       Args.append({"-target", Target});
+    if (!Stdlib.empty())
+      Args.append({"--stdlib", Stdlib});
     return Args;
   }
 
@@ -203,6 +214,7 @@ template <> struct DenseMapInfo<DriverArgs> {
         Val.Driver,
         Val.StandardIncludes,
         Val.StandardCXXIncludes,
+        Val.Stdlib,
         Val.Lang,
         Val.Sysroot,
         Val.ISysroot,

--- a/clang-tools-extra/clangd/SystemIncludeExtractor.cpp
+++ b/clang-tools-extra/clangd/SystemIncludeExtractor.cpp
@@ -214,10 +214,11 @@ template <> struct DenseMapInfo<DriverArgs> {
         Val.Driver,
         Val.StandardIncludes,
         Val.StandardCXXIncludes,
-        Val.Stdlib,
         Val.Lang,
         Val.Sysroot,
         Val.ISysroot,
+        Val.Target,
+        Val.Stdlib,
     });
   }
   static bool isEqual(const DriverArgs &LHS, const DriverArgs &RHS) {

--- a/clang-tools-extra/clangd/test/system-include-extractor.test
+++ b/clang-tools-extra/clangd/test/system-include-extractor.test
@@ -18,6 +18,7 @@
 # RUN: echo '[ -z "${args##*"--sysroot /my/sysroot/path"*}" ] || exit' >> %t.dir/bin/my_driver.sh
 # RUN: echo '[ -z "${args##*"-isysroot /isysroot"*}" ] || exit' >> %t.dir/bin/my_driver.sh
 # RUN: echo '[ -z "${args##*"-target arm-linux-gnueabihf"*}" ] || exit' >> %t.dir/bin/my_driver.sh
+# RUN: echo '[ -z "${args##*"--stdlib libc++"*}" ] || exit' >> %t.dir/bin/my_driver.sh
 # RUN: echo 'echo line to ignore >&2' >> %t.dir/bin/my_driver.sh
 # RUN: echo 'printf "Target: arm-linux-gnueabihf\r\n" >&2' >> %t.dir/bin/my_driver.sh
 # RUN: echo 'printf "#include <...> search starts here:\r\n" >&2' >> %t.dir/bin/my_driver.sh
@@ -37,7 +38,7 @@
 
 # Generate a compile_commands.json that will query the mock driver we've
 # created. Which should add a.h and b.h into include search path.
-# RUN: echo '[{"directory": "%/t.dir", "command": "my_driver.sh the-file.cpp --target=arm-linux-gnueabihf -nostdinc --sysroot /my/sysroot/path -isysroot/isysroot", "file": "the-file.cpp"}]' > %t.dir/compile_commands.json
+# RUN: echo '[{"directory": "%/t.dir", "command": "my_driver.sh the-file.cpp --target=arm-linux-gnueabihf -nostdinc --sysroot /my/sysroot/path -isysroot/isysroot -stdlib=libc++", "file": "the-file.cpp"}]' > %t.dir/compile_commands.json
 
 # RUN: sed -e "s|INPUT_DIR|%/t.dir|g" %s > %t.test.1
 # On Windows, we need the URI in didOpen to look like "uri":"file:///C:/..."
@@ -75,7 +76,7 @@
 {"jsonrpc":"2.0","method":"exit"}
 
 # Generate a different compile_commands.json which does not point to the mock driver
-# RUN: echo '[{"directory": "%/t.dir", "command": "gcc the-file.cpp --target=arm-linux-gnueabihf -nostdinc --sysroot /my/sysroot/path -isysroot/isysroot", "file": "the-file.cpp"}]' > %t.dir/compile_commands.json
+# RUN: echo '[{"directory": "%/t.dir", "command": "gcc the-file.cpp --target=arm-linux-gnueabihf -nostdinc --sysroot /my/sysroot/path -isysroot/isysroot -stdlib=libc++", "file": "the-file.cpp"}]' > %t.dir/compile_commands.json
 
 # Generate a clangd config file which points to the mock driver instead
 # RUN: echo 'CompileFlags:' > %t.dir/.clangd

